src/pkg/cluster/tunnel.go

```diff
@@ -273,3 +280,255 @@ func serviceInfoFromNodePortURL(services []corev1.Service, nodePortURL string) (
+		localPort,
+		tunnel.remotePort,
+		tunnel.resourceType,
+		tunnel.resourceName,
+		tunnel.namespace,
+	)
+	message.Debugf(msg)
+
+	// Find the pod to port forward to
+	podName, err := tunnel.getAttachablePodForResource(ctx)
+	if err != nil {
+		return "", fmt.Errorf("unable to find pod attached to given resource: %w", err)
+	}
+	message.Debugf("Selected pod %s to open port forward to", podName)
+
+	// Build url to the port forward endpoint.
+	// Example: http://localhost:8080/api/v1/namespaces/helm/pods/tiller-deploy-9itlq/portforward.
+	postEndpoint := tunnel.clientset.CoreV1().RESTClient().Post()
+	namespace := tunnel.namespace
+	portForwardCreateURL := postEndpoint.
+		Resource("pods").
+		Namespace(namespace).
+		Name(podName).
+		SubResource("portforward").
+		URL()
+
+	message.Debugf("Using URL %s to create portforward", portForwardCreateURL)
+
+	// Construct the spdy client required by the client-go portforward library.
+	transport, upgrader, err := spdy.RoundTripperFor(tunnel.restConfig)
+	if err != nil {
+		return "", fmt.Errorf("unable to create the spdy client %w", err)
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", portForwardCreateURL)
+
+	// Construct a new PortForwarder struct that manages the instructed port forward tunnel.
+	ports := []string{fmt.Sprintf("%d:%d", localPort, tunnel.remotePort)}
+	portforwarder, err := portforward.New(dialer, ports, tunnel.stopChan, tunnel.readyChan, tunnel.out, tunnel.out)
+	if err != nil {
+		return "", fmt.Errorf("unable to create the port forward: %w", err)
+	}
+
+	// Open the tunnel in a goroutine so that it is available in the background. Report errors to the main goroutine via
+	// a new channel.
+	errChan := make(chan error)
+	go func() {
+		errChan <- portforwarder.ForwardPorts()
+	}()
+
+	// Wait for an error or the tunnel to be ready.
+	select {
+	case err = <-errChan:
+		return "", fmt.Errorf("unable to start the tunnel: %w", err)
+	case <-portforwarder.Ready:
+		// Store for endpoint output
+		tunnel.localPort = localPort
+		url := tunnel.FullURL()
+
+		// Store the error channel to listen for errors
+		tunnel.errChan = errChan
+
+		message.Debugf("Creating port forwarding tunnel at %s", url)
+		return url, nil
+	}
+}
+
+// getAttachablePodForResource will find a pod that can be port forwarded to the provided resource type and return
+// the name.
+func (tunnel *Tunnel) getAttachablePodForResource(ctx context.Context) (string, error) {
+	switch tunnel.resourceType {
+	case PodResource:
+		return tunnel.resourceName, nil
+	case SvcResource:
+		return tunnel.getAttachablePodForService(ctx)
+	default:
+		return "", fmt.Errorf("unknown resource type: %s", tunnel.resourceType)
+	}
+}
+
+// getAttachablePodForService will find an active pod associated with the Service and return the pod name.
+func (tunnel *Tunnel) getAttachablePodForService(ctx context.Context) (string, error) {
```

Comment:
Taking the wait away from this function is interesting. I like that it's faster feedback to the user if the pod isn't there, however I would like to know the behavior if we return an err'd pod or a pod that isn't ready. I'm guessing the port forward command would just error?

Reply:
It would not return a failing Pod because we filter pods based on status phase. This behavior is a lot more inline with kubectl portforward. While it is nice to have some sort of waiting functionality in the tunnel I do not think it is the right place for it.

Reply:
Makes sense, agree being similar to kubectl port forward is better